botorch/models/ensemble.py

```diff
@@ -0,0 +1,95 @@
+#!/usr/bin/env python3
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+r"""
+Ensemble Models: Simple wrappers that allow the usage of ensembles
+via the BoTorch Model and Posterior APIs.
+"""
+
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import Any, List, Optional
+
+from botorch.acquisition.objective import PosteriorTransform
+from botorch.exceptions.errors import UnsupportedError
+from botorch.models.model import Model
+from botorch.posteriors.ensemble import EnsemblePosterior
+from torch import Tensor
+
+
+class EnsembleModel(Model, ABC):
+    r"""
+    Abstract base class for ensemble models.
+
+    :meta private:
+    """
+
+    @abstractmethod
+    def forward(self, X: Tensor) -> Tensor:
+        r"""Compute the (ensemble) model output at X.
+
+        Args:
+            X: A `batch_shape x n x d`-dim input tensor `X`.
+
+        Returns:
+            A `batch_shape x s x n x m`-dimensional output tensor where
+            `s` is the size of the ensemble.
+        """
+        pass  # pragma: no cover
+
+    def _forward(self, X: Tensor) -> Tensor:
+        return self.forward(X=X)
+
+    @property
+    def num_outputs(self) -> int:
+        r"""The number of outputs of the model."""
+        return self._num_outputs
+
+    def posterior(
+        self,
+        X: Tensor,
+        output_indices: Optional[List[int]] = None,
+        posterior_transform: Optional[PosteriorTransform] = None,
+        **kwargs: Any,
+    ) -> EnsemblePosterior:
+        r"""Compute the ensemble posterior at X.
+
+        Args:
+            X: A `batch_shape x n x d`-dim input tensor `X`.
```

Comment:
For consistency, we refer to -2 dim of the model input as q-batch dim.
```suggestion
            X: A `batch_shape x q x d`-dim input tensor `X`.
```